msg.go

```diff
@@ -962,70 +962,76 @@ func compressedLen(dns *Msg, compress bool) int {
 	return l
 }
 
-func compressionLenSlice(len int, c map[string]int, rs []RR) int {
-	var l int
+func compressionLenSlice(lenp int, c map[string]int, rs []RR) int {
+	initLen := lenp
 	for _, r := range rs {
 		if r == nil {
 			continue
 		}
-		// track this length, and the global length in len, while taking compression into account for both.
+		// TmpLen is to track len of record at 14bits boudaries, 6 bytes is the raw overhead
+		tmpLen := lenp + 6
 		x := r.len()
-		l += x
-		len += x
-
-		k, ok := compressionLenSearch(c, r.Header().Name)
+		// track this length, and the global length in len, while taking compression into account for both.
+		k, ok, sz := compressionLenSearch(c, r.Header().Name)
 		if ok {
-			l += 1 - k
-			len += 1 - k
-		}
-
-		if len < maxCompressionOffset {
-			compressionLenHelper(c, r.Header().Name)
+			x += 1 - k
 		}
+		tmpLen += sz
 
-		k, ok = compressionLenSearchType(c, r)
+		tmpLen += compressionLenHelper(c, r.Header().Name, tmpLen)
+		k, ok, added := compressionLenSearchType(c, r)
 		if ok {
-			l += 1 - k
-			len += 1 - k
-		}
-
-		if len < maxCompressionOffset {
-			compressionLenHelperType(c, r)
+			x += 1 - k
 		}
+		tmpLen += added
+		tmpLen += compressionLenHelperType(c, r, tmpLen)
+		lenp += x
 	}
-	return l
+	return lenp - initLen
 }
 
-// Put the parts of the name in the compression map.
-func compressionLenHelper(c map[string]int, s string) {
+// Put the parts of the name in the compression map, return the size added in payload
+func compressionLenHelper(c map[string]int, s string, currentLen int) int {
+	addedSize := 0
 	pref := ""
 	lbs := Split(s)
 	for j := len(lbs) - 1; j >= 0; j-- {
 		pref = s[lbs[j]:]
 		if _, ok := c[pref]; !ok {
-			c[pref] = len(pref)
+			lenAdded := len(pref)
+			numLabelsBefore := len(lbs) - j - 1
+			offsetOfLabel := currentLen + len(s) - len(pref) + numLabelsBefore*2
```

Comment:
don't fully understand this yet... why the *2 ?

Reply:
numLabelsBefore*2 => size of label pointers (since their size is 2 bytes => 2 times)